massa-versioning/src/versioning.rs

```diff
@@ -872,74 +879,107 @@ impl MipStoreRaw {
     fn advance_states_on_updated_stats(&mut self, slot_timestamp: MassaTime) {
+        let mut reset_stats = false;
+
         for (mi, state) in self.store.iter_mut() {
+            if state.is_final() {
+                // State cannot change (ex: Active), no need to update
+                continue;
+            }
+
             let network_version_count = *self
                 .stats
                 .network_version_counters
                 .get(&mi.version)
-                .unwrap_or(&0) as f32;
-            let block_count_considered = self.stats.config.block_count_considered as f32;
-
-            let vote_ratio_ = 100.0 * network_version_count / block_count_considered;
-
-            let vote_ratio = Amount::const_init(vote_ratio_.round() as u64, 0);
-
-            debug!("[VERSIONING STATS] vote_ratio = {} (from version counter = {} and blocks considered = {})", vote_ratio, network_version_count, block_count_considered);
+                .unwrap_or(&0);
+            let block_count_considered = self.stats.config.block_count_considered;
+            let vote_ratio_ = Amount::from_mantissa_scale(network_version_count, 0).map(|a| {
+                a.saturating_mul_u64(100)
+                    .checked_div_u64(u64::try_from(block_count_considered).unwrap_or(0))
+            });
+
+            if let Ok(Some(vote_ratio)) = vote_ratio_ {
+                debug!("[VERSIONING STATS] vote ratio = {} (network version counter = {} - blocks considered = {})", vote_ratio, network_version_count, block_count_considered);
+
+                let advance_msg = Advance {
+                    start_timestamp: mi.start,
+                    timeout: mi.timeout,
+                    threshold: vote_ratio,
+                    now: slot_timestamp,
+                    activation_delay: mi.activation_delay,
+                };
+
+                state.on_advance(&advance_msg.clone());
+
+                if vote_ratio > VERSIONING_THRESHOLD_TRANSITION_ACCEPTED {
+                    // A new version is going to be 'Active' - reset stats
+                    reset_stats = true;
+                }
+            }
+        }
 
-            let advance_msg = Advance {
-                start_timestamp: mi.start,
-                timeout: mi.timeout,
-                threshold: vote_ratio,
-                now: slot_timestamp,
-                activation_delay: mi.activation_delay,
-            };
+        if reset_stats {
+            // Reset stats after loop as we can have 2 MIP
+            // One become LockedIn & One become Started
+            // FIXME: is this true? Can we have LockedIn & Started at the same time
+            //        can have LockedIn & Defined at the same time
 
-            // TODO / OPTIM: filter the store to avoid advancing on failed and active versions
-            state.on_advance(&advance_msg.clone());
+            debug!("[VERSIONING STATS] reset stats (threshold reached)");
+            self.stats.reset();
```

Comment:
Is this needed or just a clean up?
We have to check exactly what we're doing, but I think if we are in LockedIn we still announce the version in our blocks. This is useful because we can see if other people have updated between the lockedIn and active.

For example: MIP_0001 reaches 75% of blocks and becomes lockedIn, and we don't reset the stats. We can follow the counter and see that 90% of blocks vote for that version just before activation. So we know we fork with only 10% of the network

Reply:
Removed the reset. Now we have as much counter as block considered + remove the counter when its value is 0...